### PR TITLE
weka: use openjdk11

### LINKED ISSUE #199741: Weka crashes on startup with Java 17
### Describe the bug
Weka crashes if using the _nixpkgs_ version

### Steps To Reproduce
1. Get weka from nixpkgs at f994293d1eb8812f032e8919e10a594567cf6ef7 (nixos-unstable)
2. Open weka

### Expected behavior
Weka opening that main window and working properly.

It neither opens the main window nor works properly.

### Additional context
Downloading the zip from [Weka's official site](https://sourceforge.net/projects/weka/) and running by using `java -jar weka.jar` from `openjdk11` works as expected.

```
lucasew@riverwood ~/WORKSPACE/TCC 0$ /nix/store/mhh9vz3rw8dj813y0q9fxykxpv7bykyi-weka-3.9.2/bin/weka 
java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @1c8a1bf1
	java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	weka.core.WekaPackageClassLoaderManager.injectClasses(WekaPackageClassLoaderManager.java:790)
	weka.core.WekaPackageClassLoaderManager.injectAllClassesInFromStream(WekaPackageClassLoaderManager.java:688)
	weka.core.WekaPackageClassLoaderManager.injectMTJCoreClasses(WekaPackageClassLoaderManager.java:118)
	weka.core.WekaPackageManager.<clinit>(WekaPackageManager.java:241)
	weka.core.ResourceUtils.readProperties(ResourceUtils.java:241)
	weka.core.ResourceUtils.readProperties(ResourceUtils.java:184)
	weka.core.Utils.readProperties(Utils.java:151)
	weka.core.logging.Logger.<clinit>(Logger.java:50)
	weka.gui.GUIChooserApp.main(GUIChooserApp.java:1607)
	java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.base/java.lang.reflect.Method.invoke(Method.java:568)
	weka.gui.SplashWindow.invokeMain(SplashWindow.java:306)
	weka.gui.GUIChooser.main(GUIChooser.java:92)

	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at weka.core.WekaPackageClassLoaderManager.injectClasses(WekaPackageClassLoaderManager.java:790)
	at weka.core.WekaPackageClassLoaderManager.injectAllClassesInFromStream(WekaPackageClassLoaderManager.java:688)
	at weka.core.WekaPackageClassLoaderManager.injectMTJCoreClasses(WekaPackageClassLoaderManager.java:118)
	at weka.core.WekaPackageManager.<clinit>(WekaPackageManager.java:241)
	at weka.core.ResourceUtils.readProperties(ResourceUtils.java:241)
	at weka.core.ResourceUtils.readProperties(ResourceUtils.java:184)
	at weka.core.Utils.readProperties(Utils.java:151)
	at weka.core.logging.Logger.<clinit>(Logger.java:50)
	at weka.gui.GUIChooserApp.main(GUIChooserApp.java:1607)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at weka.gui.SplashWindow.invokeMain(SplashWindow.java:306)
	at weka.gui.GUIChooser.main(GUIChooser.java:92)
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by weka.gui.GUIChooserApp (file:/nix/store/mhh9vz3rw8dj813y0q9fxykxpv7bykyi-weka-3.9.2/share/weka/weka.jar)
WARNING: Please consider reporting this to the maintainers of weka.gui.GUIChooserApp
WARNING: System::setSecurityManager will be removed in a future release
Exception in thread "main" java.lang.InternalError: Failed to invoke main method
	weka.gui.SplashWindow.invokeMain(SplashWindow.java:308)
	weka.gui.GUIChooser.main(GUIChooser.java:92)

	at weka.gui.SplashWindow.invokeMain(SplashWindow.java:308)
	at weka.gui.GUIChooser.main(GUIChooser.java:92)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at weka.gui.SplashWindow.invokeMain(SplashWindow.java:306)
	... 1 more
Caused by: java.lang.NoClassDefFoundError: no/uib/cipr/matrix/Matrix
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:467)
	at weka.core.WekaPackageClassLoaderManager.forName(WekaPackageClassLoaderManager.java:198)
	at weka.core.WekaPackageClassLoaderManager.forName(WekaPackageClassLoaderManager.java:178)
	at weka.core.ClassDiscovery.find(ClassDiscovery.java:247)
	at weka.gui.GenericPropertiesCreator.generateOutputProperties(GenericPropertiesCreator.java:542)
	at weka.gui.GenericPropertiesCreator.execute(GenericPropertiesCreator.java:639)
	at weka.gui.GenericPropertiesCreator.<clinit>(GenericPropertiesCreator.java:167)
	at weka.core.WekaPackageManager.refreshGOEProperties(WekaPackageManager.java:1403)
	at weka.core.WekaPackageManager.loadPackages(WekaPackageManager.java:1328)
	at weka.core.WekaPackageManager.loadPackages(WekaPackageManager.java:1205)
	at weka.gui.GenericObjectEditor.determineClasses(GenericObjectEditor.java:172)
	at weka.gui.GenericObjectEditor.<clinit>(GenericObjectEditor.java:242)
	at weka.gui.GUIChooserApp.<init>(GUIChooserApp.java:662)
	at weka.gui.GUIChooserApp.createSingleton(GUIChooserApp.java:176)
	at weka.gui.GUIChooserApp.main(GUIChooserApp.java:1676)
	... 6 more
Caused by: java.lang.ClassNotFoundException: no.uib.cipr.matrix.Matrix
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 22 more
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### PR DESCRIPTION
**Pull request by lucasew**
Signed-off-by: lucasew <lucas59356@gmail.com>

###### Description of changes
Use Java 11 instead of 17 to run Weka

Closes #199741

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lucasew** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>weka</li>
  </ul>
</details>

**Review by thiagokokada — COMMENTED** [review]
Upstream seems to use JDK17: https://waikato.github.io/weka-wiki/downloading_weka/

File: pkgs/applications/science/math/weka/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, jre, unzip, makeWrapper }:
+{ lib, stdenv, fetchurl, jdk11, unzip, makeWrapper }:
```
Comment (AndersonTorres) [inline pkgs/applications/science/math/weka/default.nix:1]:
```suggestion
{ lib, stdenv, fetchurl, jdk, unzip, makeWrapper }:
```

File: pkgs/applications/science/math/weka/default.nix [hunk 2]
```diff
@@ -17,7 +17,7 @@ stdenv.mkDerivation rec {
     mkdir -pv $out/share/weka
     cp -Rv * $out/share/weka
 
-    makeWrapper ${jre}/bin/java $out/bin/weka \
+    makeWrapper ${jdk11}/bin/java $out/bin/weka \
```
Comment (AndersonTorres) [inline pkgs/applications/science/math/weka/default.nix:20]:
```suggestion
    makeWrapper ${jdk}/bin/java $out/bin/weka \
```

Parameterize the version in `all-packages.nix`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by lucasew** [comment]
> Upstream seems to use JDK17: https://waikato.github.io/weka-wiki/downloading_weka/

They also promote an alternative JDK based on OpenJDK (Azul Zulu).

But for some reason, OpenJDK 17 crashes at the startup and OpenJDK 11 is working nicely.

**Comment by thiagokokada** [comment]
> They also promote an alternative JDK based on OpenJDK (Azul Zulu).

Yeah, I saw this too. We have Zulu packaged in nixpkgs too: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/zulu/default.nix. However I wouldn't recommend targeting it as default.

> But for some reason, OpenJDK 17 crashes at the startup and OpenJDK 11 is working nicely.

Both Zulu and OpenJDK need to pass the same test suite, and technically they should be 100% compatible with each other. So I am thinking the issue you're having is related with some flag that is missing on the wrapper or something else.

I would prefer we investigate this first instead of changing the default JDK to 11.

**Comment by thiagokokada** [comment]
I downloaded the `weka-3-9-6-azul-zulu-linux.zip` file. It includes a `weka.sh` script, that pass a bunch of `--add-opens` flags that we should replicate here, and probably will fix the issue you're seeing.

I would prefer that we package this script or replicate the flags inside the `makeWrapper` argument, since this is more aligned with upstream. However, if the maintainer (@mimame) thinks the current approach is fine I am also fine.

**Comment by AndersonTorres** [comment]
For now I prefer to merge this PR.

**Pull request merged**